### PR TITLE
Update for PR #457

### PR DESCRIPTION
**Pull request by randombit**
Addresses some issues in PR #457 I noticed.

### CODE REVIEW

**Comment by randombit** [comment]
@mgierlings @DanielTBrown I would like your feedback on my modifications to the TLS::Callbacks class in 487832e

With my changes on top of PR 457, users of the 'old' TLS API (ie how callbacks have worked from 1.11.0 to the current release) will not have to make any changes to their code at first. Previously the alert callback changed even for the deprecated compat constructor. Instead applications opt into the API changes by deriving from TLS::Callbacks directly. It also removes some (probably minimal) overhead of carrying empty std::functions, when the application is using the new Callbacks class to set the application actions.

Right now I plan on merging this PR at 7a83b81 sometime in the coming week.

**Comment by mgierlings** [comment]
@randombit Thank you, the changes look good to me. I especially like the new names you introduced for the callback methods and the extraction of the legacy callbacks to the Compat_Callbacks class.

**Comment by randombit** [comment]
I realized the server still had a channel-specific callback which was the ALPN callback. Moved that to TLS::Callbacks in dd5cda3 and squashed some of the prior commits to reduce log noise.

**Comment by DanielTBrown** [comment]
@randombit Thank you, the new callback class looks very nice, and easy to use. If you would like doxygen function parameter comments, I have posted PR 595. The pull request contains no code changes, only comment updates.

File: src/lib/tls/tls_server.cpp [hunk 1]
```diff
@@ -284,441 +310,516 @@ Server::get_peer_cert_chain(const Handshake_State& state) const
...
-      state.compute_session_keys();
-      }
-   else if(type == CERTIFICATE_VERIFY)
-      {
-      state.client_verify(new Certificate_Verify(contents, state.version()));
+      pending_state.server_finished(new Finished(pending_state.handshake_io(), pending_state, SERVER));
+      pending_state.set_expected_next(HANDSHAKE_CCS);
+   }
+
+void Server::session_create(Server_Handshake_State& pending_state,
```
Comment (SonarQubeAnalysis) [inline src/lib/tls/tls_server.cpp]: MAJOR This function/method has 116 lines, which is greater than the 100 lines authorized. Split it into smaller functions. [rule](https://sonarqube.com/coding_rules#rule_key=cpp%3AS138)

File: src/lib/tls/tls_server.cpp [hunk 2]
```diff
@@ -284,441 +310,516 @@ Server::get_peer_cert_chain(const Handshake_State& state) const
...
+      case HANDSHAKE_CCS:
+         this->process_change_cipher_spec_msg(state);
+      break;
+
+      case FINISHED:
+         this->process_finished_msg(state, type, contents);
+      break;
+
+      default:
+         throw Unexpected_Message("Unknown handshake message received");
```
Comment (SonarQubeAnalysis) [inline src/lib/tls/tls_server.cpp]: MAJOR Remove the code after this "throw". [rule](https://sonarqube.com/coding_rules#rule_key=cpp%3AS1763)

File: src/cli/tls_client.cpp [hunk 3]
```diff
@@ -312,16 +305,19 @@ class TLS_Client final : public Command
...
          }
 
-      void process_data(const uint8_t buf[], size_t buf_size)
+      void tls_record_received(uint64_t seq_no, const uint8_t buf[], size_t buf_size)
          {
          for(size_t i = 0; i != buf_size; ++i)
             output() << buf[i];
          }
+
+      private:
```
Comment (SonarQubeAnalysis) [inline src/cli/tls_client.cpp]: MINOR Remove this redundant access specifier; it does not change the accessibility level. [rule](https://sonarqube.com/coding_rules#rule_key=cpp%3AS3539)

File: src/lib/tls/tls_client.cpp [hunk 4]
```diff
@@ -42,6 +43,23 @@ class Client_Handshake_State : public Handshake_State
 /*
 * TLS Client Constructor
 */
+Client::Client(Callbacks& callbacks,
```
Comment (SonarQubeAnalysis) [inline src/lib/tls/tls_client.cpp]: MAJOR This function has 9 parameters, which is greater than the 7 authorized. [rule](https://sonarqube.com/coding_rules#rule_key=cpp%3AS107)

File: src/lib/tls/tls_blocking.cpp [hunk 5]
```diff
@@ -35,14 +39,15 @@ Blocking_Client::Blocking_Client(read_fn reader,
              offer_version,
              next)
    {
+   printf("hi\n");
```
Comment (SonarQubeAnalysis) [inline src/lib/tls/tls_blocking.cpp]: MAJOR Replace this usage of "printf" by a logger. [rule](https://sonarqube.com/coding_rules#rule_key=cpp%3AS106)

File: src/tests/unit_tls.cpp [hunk 6]
```diff
@@ -444,183 +501,234 @@ Test::Result test_dtls_handshake(Botan::TLS::Protocol_Version offer_version,
...
 
-            // If we corrupted a DTLS application message, resend it:
-            if(client.is_active() && corrupt_client_data && server_recv.empty())
-               client.send(client_sent);
-            if(server.is_active() && corrupt_server_data && client_recv.empty())
-               server.send(server_sent);
+               // If we corrupted a DTLS application message, resend it:
+               if(client->is_active() && corrupt_client_data && server_recv.empty())
+                  client->send(client_sent);
+               if(server->is_active() && corrupt_server_data && client_recv.empty())
```
Comment (SonarQubeAnalysis) [inline src/tests/unit_tls.cpp]: MAJOR Refactor this code to not nest more than 3 if|for|do|while|switch statements. [rule](https://sonarqube.com/coding_rules#rule_key=cpp%3AS134)

File: src/tests/unit_tls.cpp [hunk 7]
```diff
@@ -444,183 +501,234 @@ Test::Result test_dtls_handshake(Botan::TLS::Protocol_Version offer_version,
...
+                  server->send(server_sent);
 
-            if(client_recv.size())
-               {
-               result.test_eq("client recv", client_recv, server_sent);
-               }
+               if(client_recv.size())
```
Comment (SonarQubeAnalysis) [inline src/tests/unit_tls.cpp]: MAJOR Refactor this code to not nest more than 3 if|for|do|while|switch statements. [rule](https://sonarqube.com/coding_rules#rule_key=cpp%3AS134)

File: src/tests/unit_tls.cpp [hunk 8]
```diff
@@ -218,164 +223,216 @@ Test::Result test_tls_handshake(Botan::TLS::Protocol_Version offer_version,
...
+                  if(client_recv.size() && server_recv.size())
+                     {
+                     result.test_failure("Negotiated in the face of data corruption " + std::to_string(r));
+                     }
                   }
-               }
 
-            if(client.is_closed() && server.is_closed())
-               break;
+               if(client->is_closed() && server->is_closed())
```
Comment (SonarQubeAnalysis) [inline src/tests/unit_tls.cpp]: MAJOR Refactor this code to not nest more than 3 if|for|do|while|switch statements. [rule](https://sonarqube.com/coding_rules#rule_key=cpp%3AS134)

File: src/tests/unit_tls.cpp [hunk 9]
```diff
@@ -161,7 +162,11 @@ std::function<void (const byte[], size_t)> queue_inserter(std::vector<byte>& q)
    return [&](const byte buf[], size_t sz) { q.insert(q.end(), buf, buf + sz); };
    }
 
-void print_alert(Botan::TLS::Alert, const byte[], size_t)
+void print_alert(Botan::TLS::Alert)
+   {
```
Comment (SonarQubeAnalysis) [inline src/tests/unit_tls.cpp]: MAJOR Add a nested comment explaining why this method is empty, or complete the implementation. [rule](https://sonarqube.com/coding_rules#rule_key=cpp%3AS1186)

File: src/tests/unit_tls.cpp [hunk 10]
```diff
@@ -218,164 +223,216 @@ Test::Result test_tls_handshake(Botan::TLS::Protocol_Version offer_version,
...
-            if(client.is_closed() && server.is_closed())
-               break;
+               if(client->is_closed() && server->is_closed())
+                  break;
 
-            if(server_recv.size() && client_recv.size())
-               {
-               Botan::SymmetricKey client_key = client.key_material_export("label", "context", 32);
-               Botan::SymmetricKey server_key = server.key_material_export("label", "context", 32);
+               if(server_recv.size() && client_recv.size())
```
Comment (SonarQubeAnalysis) [inline src/tests/unit_tls.cpp]: MAJOR Refactor this code to not nest more than 3 if|for|do|while|switch statements. [rule](https://sonarqube.com/coding_rules#rule_key=cpp%3AS134)

File: src/tests/unit_tls.cpp [hunk 11]
```diff
@@ -444,183 +501,234 @@ Test::Result test_dtls_handshake(Botan::TLS::Protocol_Version offer_version,
...
+               if(client_recv.size())
+                  {
+                  result.test_eq("client recv", client_recv, server_sent);
+                  }
 
-            if(server_recv.size())
-               {
-               result.test_eq("server recv", server_recv, client_sent);
-               }
+               if(server_recv.size())
```
Comment (SonarQubeAnalysis) [inline src/tests/unit_tls.cpp]: MAJOR Refactor this code to not nest more than 3 if|for|do|while|switch statements. [rule](https://sonarqube.com/coding_rules#rule_key=cpp%3AS134)

File: src/tests/unit_tls.cpp [hunk 12]
```diff
@@ -444,183 +501,234 @@ Test::Result test_dtls_handshake(Botan::TLS::Protocol_Version offer_version,
...
+                  {
+                  result.test_eq("server recv", server_recv, client_sent);
+                  }
 
-            if(client.is_closed() && server.is_closed())
-               break;
+               if(client->is_closed() && server->is_closed())
```
Comment (SonarQubeAnalysis) [inline src/tests/unit_tls.cpp]: MAJOR Refactor this code to not nest more than 3 if|for|do|while|switch statements. [rule](https://sonarqube.com/coding_rules#rule_key=cpp%3AS134)

File: src/tests/unit_tls.cpp [hunk 13]
```diff
@@ -444,183 +501,234 @@ Test::Result test_dtls_handshake(Botan::TLS::Protocol_Version offer_version,
...
+                  break;
 
-            if(server_recv.size() && client_recv.size())
-               {
-               Botan::SymmetricKey client_key = client.key_material_export("label", "context", 32);
-               Botan::SymmetricKey server_key = server.key_material_export("label", "context", 32);
+               if(server_recv.size() && client_recv.size())
```
Comment (SonarQubeAnalysis) [inline src/tests/unit_tls.cpp]: MAJOR Refactor this code to not nest more than 3 if|for|do|while|switch statements. [rule](https://sonarqube.com/coding_rules#rule_key=cpp%3AS134)

File: src/tests/unit_tls.cpp [hunk 14]
```diff
@@ -444,183 +501,234 @@ Test::Result test_dtls_handshake(Botan::TLS::Protocol_Version offer_version,
...
-               /*
-               * Use this as a temp value to hold the queues as otherwise they
-               * might end up appending more in response to messages during the
-               * handshake.
-               */
-               std::vector<byte> input;
-               std::swap(c2s_traffic, input);
-
-               if(corrupt_server_data)
+               if(client->is_active() && client_sent.empty())
```
Comment (SonarQubeAnalysis) [inline src/tests/unit_tls.cpp]: MAJOR Refactor this code to not nest more than 3 if|for|do|while|switch statements. [rule](https://sonarqube.com/coding_rules#rule_key=cpp%3AS134)

File: src/tests/unit_tls.cpp [hunk 15]
```diff
@@ -218,164 +223,216 @@ Test::Result test_tls_handshake(Botan::TLS::Protocol_Version offer_version,
...
-                     total_consumed += needed;
+                     size_t needed = client->received_data(input.data(), input.size());
+                     result.test_eq("full packet received", needed, 0);
                      }
+
+                  continue;
                   }
-               else
+
+               if(client_recv.size())
```
Comment (SonarQubeAnalysis) [inline src/tests/unit_tls.cpp]: MAJOR Refactor this code to not nest more than 3 if|for|do|while|switch statements. [rule](https://sonarqube.com/coding_rules#rule_key=cpp%3AS134)

File: src/tests/unit_tls.cpp [hunk 16]
```diff
@@ -444,183 +501,234 @@ Test::Result test_dtls_handshake(Botan::TLS::Protocol_Version offer_version,
...
+                        {
+                        result.test_failure("server error", e.what());
+                        }
                      }
+
+                  continue;
                   }
-               else
+
+               if(s2c_traffic.size() > 0)
```
Comment (SonarQubeAnalysis) [inline src/tests/unit_tls.cpp]: MAJOR Refactor this code to not nest more than 3 if|for|do|while|switch statements. [rule](https://sonarqube.com/coding_rules#rule_key=cpp%3AS134)

File: src/tests/unit_tls.cpp [hunk 17]
```diff
@@ -218,164 +223,216 @@ Test::Result test_tls_handshake(Botan::TLS::Protocol_Version offer_version,
...
-            if(client_recv.size())
-               {
-               result.test_eq("client recv", client_recv, server_sent);
-               }
-
-            if(server_recv.size())
-               {
-               result.test_eq("server recv", server_recv, client_sent);
-               }
+               if(server_recv.size())
```
Comment (SonarQubeAnalysis) [inline src/tests/unit_tls.cpp]: MAJOR Refactor this code to not nest more than 3 if|for|do|while|switch statements. [rule](https://sonarqube.com/coding_rules#rule_key=cpp%3AS134)

File: src/tests/unit_tls.cpp [hunk 18]
```diff
@@ -218,164 +223,216 @@ Test::Result test_tls_handshake(Botan::TLS::Protocol_Version offer_version,
...
-               {
-               result.test_eq("server protocol", server.next_protocol(), "test/3");
+               if(handshake_done && (client->is_closed() || server->is_closed()))
+                  break;
 
-               const size_t s_len = 1 + rng.next_byte() + rng.next_byte();
-               server_sent = unlock(rng.random_vec(s_len));
-               server.send(server_sent);
-               }
+               if(client->is_active() && client_sent.empty())
```
Comment (SonarQubeAnalysis) [inline src/tests/unit_tls.cpp]: MAJOR Refactor this code to not nest more than 3 if|for|do|while|switch statements. [rule](https://sonarqube.com/coding_rules#rule_key=cpp%3AS134)

File: src/tests/unit_tls.cpp [hunk 19]
```diff
@@ -444,183 +501,234 @@ Test::Result test_dtls_handshake(Botan::TLS::Protocol_Version offer_version,
...
-                  catch(std::exception&)
-                     {
-                     result.test_note("corruption caused server exception");
-                     }
+                  // TODO send multiple parts
+                  client->send(client_sent);
                   }
-               else
+
+               if(server->is_active() && server_sent.empty())
```
Comment (SonarQubeAnalysis) [inline src/tests/unit_tls.cpp]: MAJOR Refactor this code to not nest more than 3 if|for|do|while|switch statements. [rule](https://sonarqube.com/coding_rules#rule_key=cpp%3AS134)

File: src/tests/unit_tls.cpp [hunk 20]
```diff
@@ -218,164 +223,216 @@ Test::Result test_tls_handshake(Botan::TLS::Protocol_Version offer_version,
...
-               }
+               if(server_recv.size())
+                  {
+                  result.test_eq("server->recv", server_recv, client_sent);
+                  }
 
-            if(r > 2)
-               {
-               if(client_recv.size() && server_recv.size())
+               if(r > 2)
```
Comment (SonarQubeAnalysis) [inline src/tests/unit_tls.cpp]: MAJOR Refactor this code to not nest more than 3 if|for|do|while|switch statements. [rule](https://sonarqube.com/coding_rules#rule_key=cpp%3AS134)

File: src/tests/unit_tls.cpp [hunk 21]
```diff
@@ -444,183 +501,234 @@ Test::Result test_dtls_handshake(Botan::TLS::Protocol_Version offer_version,
...
+                  continue;
+                  }
 
-            // If we corrupted a DTLS application message, resend it:
-            if(client.is_active() && corrupt_client_data && server_recv.empty())
-               client.send(client_sent);
-            if(server.is_active() && corrupt_server_data && client_recv.empty())
-               server.send(server_sent);
+               // If we corrupted a DTLS application message, resend it:
+               if(client->is_active() && corrupt_client_data && server_recv.empty())
```
Comment (SonarQubeAnalysis) [inline src/tests/unit_tls.cpp]: MAJOR Refactor this code to not nest more than 3 if|for|do|while|switch statements. [rule](https://sonarqube.com/coding_rules#rule_key=cpp%3AS134)

**Comment by SonarQubeAnalysis** [comment]
SonarQube analysis reported 22 issues
* MAJOR 21 major
* MINOR 1 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [tls_client.cpp#L278](https://github.com/randombit/botan/blob/ee60a29088fc6dd712c1651af1e7f56a26f40d63/src/cli/tls_client.cpp#L278): Remove this unused private function. [rule](https://sonarqube.com/coding_rules#rule_key=cpp%3AS1144)

**Comment by mgierlings** [comment]
@randombit Very nice, thanks for the hard work!

**Pull request merged**